fuzzers/main.py

```diff
@@ -0,0 +1,949 @@
+            log_fullpath: the absolute path to the log file
+        """
+
+        self.corpus_vers = corpus_vers
+        self.branch = branch
+        self.target = target
+        self.repo_dir = repo_dir
+        self.log_relpath = log_relpath
+        self.log_fullpath = log_fullpath
+        self.log_file = open(log_fullpath, 'a', encoding='utf-8')  # pylint: disable=consider-using-with
+
+        self.last_time = 0.
+        self.time_paused = None
+        self.proc: typing.Any = None  # There's some weirdness around brackets and Popen
+
+        self.fuzz_build_time_metric = FUZZ_BUILD_TIME.labels(
+            branch['name'], target['crate'], target['runner'])
+        self.fuzz_time_metric = FUZZ_TIME.labels(branch['name'],
+                                                 target['crate'],
+                                                 target['runner'],
+                                                 target['flags'])
+        self.fuzz_crashes_metric = FUZZ_CRASHES.labels(branch['name'],
+                                                       target['crate'],
+                                                       target['runner'],
+                                                       target['flags'])
+
+    def build(self) -> None:
+        """
+        Build the fuzzer runner.
+
+        This is a synchronous operation for convenience reasons, as cargo itself uses all the
+        available cores most of the time it shouldn't be a big deal. The only drawback is that
+        requests to pause/exit the fuzzer would block until the current build is completed.
+        """
+        print(f'Building fuzzer for branch {self.branch} and target '
+              f'{self.target}, log is at {self.log_relpath}', file=sys.stderr)
+
+        # Log metadata information
+        current_commit = subprocess.check_output(
+            ('git', 'rev-parse', 'HEAD'),
+            cwd=self.repo_dir,
+            encoding='utf-8').stdout.strip()
+        self.log_file.write(f'Corpus version: {self.corpus_vers}\n')
+        self.log_file.write(
+            f'On commit {current_commit} (tip of branch {self.branch["name"]})\n'
+        )
+        self.log_file.write(f'Target is {self.target}\n')
+        self.log_file.write(f'Current time: {datetime.datetime.utcnow()}\n')
+        self.log_file.write(f'On host: {socket.gethostname()}\n')
+        self.log_file.flush()
+
+        # Build the fuzzer runner
+        # TODO: actually build and have build time metrics for the fuzzer runner, but this must not
+        # block the pause/resume behavior
+        #build_start = time.monotonic()
+        #subprocess.check_call(
+        #    ['cargo', 'fuzz', 'build', self.target['runner']],
+        #    cwd=self.repo_dir / self.target['crate'],
+        #    stdout=self.log_file,
+        #    stderr=subprocess.STDOUT,
+        #)
+        #self.fuzz_build_time_metric.inc(time.monotonic() - build_start)
```

Comment:
Maybe for now we can `('nice', '-n20', 'ionice', '-tn7', 'cargo', 'fuzz', 'build', self.target['runner'])` and hope it won’t interfere with NayDuck running tests too much?

Reply:
So for now the build just happens in the `cargo fuzz run`, which means it gets paused alongside the rest of the fuzzing if a nayduck test happens during building.

The reason why I'd like to add a separate build step is just in order to have metrics about build time alongside the metrics for corpus syncing time and the metrics for fuzzing time so as to fine-tune the parameters.

So I'm not sure this benefit is worth the risk that it interferes with NayDuck running tests, until I can actually fix that up properly and have it paused as it should